### PR TITLE
Order status Mark IV: manual status change

### LINKED ISSUE #121: Order Details: Manual Status Change
### Details:
@ctarda there are probably a bunch of steps missing here. I've created a rough outline. If you get stuck along the way, please ask questions :)

Internal Ref: p99K0U-Zh-p2 

The Orders i7 Designs have updated the Manual Status Change screen. The new screen has been added below.

#### Networking
- [ ] Create a new function that makes an API request in `OrdersRemote` (See also [Update an Order](https://woocommerce.github.io/woocommerce-rest-api-docs/#update-an-order))
- [ ] Determine if the success response returns an order object (if so, you can use that to update the order status)
- [ ] Update the order in Storage (probably by using `OrderStore.swift` `updateOrder(siteID: orderID: statusKey:)`)

#### Yosemite
- [ ] Create an "Update Status" Action in `OrderStatusAction`

#### UI
- [ ] The `Summary` TableViewCell displays an icon in it's bottom right corner
- [ ] Use the `statusResultsController` to list all statuses in a separate view controller
- [ ] Dispatch the action upon tapping the "Set" button (in the view controller)

screen shot 2019-03-05 at 9 44 31 am

manual-status-change-screen

Ref: #249 

### Why is this post-MVLP?
**Context** p99K0U-C7-p2

### PR DESCRIPTION
**Pull request by ctarda**
Closes #121 

This PR builds upon @mindgraffiti 's work in  #732, and won't be ready for a proper review until #732 is merged. But in the meantime, any feedback would be more than welcome! Thanks @mindgraffiti for your work on the BL and data transfer layers, integrating the UI was super easy! 

This is how it looks in this first iteration:

screen recording 2019-03-06 at 11 05 54 am mov

cc @kyleaparker 

## Testing
- Checkout the branch, run the tests.
- Take a good look at the code.
- For extra credit, make sure your store supports custom order statuses.
- Navigate to your order list, select one order, tap the pencil button
- Change the order status
- Check that the order actually changes status in the store backend, and the new status is reflected in the Order Details and Order List views
- Change the status of an order again. Tap the "Undo" button in the Notice. The change should get undone.
- Check that the relevant Tracks events are sent when tapping the pencil button and when tapping Apply in the Order Status List.

Update release notes:

- [ ] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

### CODE REVIEW

**Comment by ctarda** [comment]
@bummytime @mindgraffiti I think this is ready for a proper review now. I know you folks are busy, but if you could take a look, that would be awesome. Thanks!

File: WooCommerce/Classes/ViewRelated/Orders/Cells/SummaryTableViewCell.swift [hunk 1]
```diff
@@ -1,10 +1,11 @@
 import UIKit
 import Yosemite
+import Gridicons
 
 
 // MARK: - SummaryTableViewCell
 //
-class SummaryTableViewCell: UITableViewCell {
+final class SummaryTableViewCell: UITableViewCell {
```
Comment (bummytime) [inline WooCommerce/Classes/ViewRelated/Orders/Cells/SummaryTableViewCell.swift]: 🎉 😄

File: WooCommerce/Classes/ViewRelated/Orders/OrderDetailsViewController.swift [hunk 2]
```diff
@@ -36,6 +36,16 @@ class OrderDetailsViewController: UIViewController {
         return EntityListener(storageManager: AppDelegate.shared.storageManager, readOnlyEntity: viewModel.order)
     }()
 
+    /// ResultsController: Surrounds us. Binds the galaxy together. And also, keeps the UITableView <> (Stored) OrderStatuses in sync.
```
Comment (bummytime) [inline WooCommerce/Classes/ViewRelated/Orders/OrderDetailsViewController.swift]: @jleandroperez's `/// ResultsController: Surrounds us. Binds the galaxy together...` still lives on ❤️

File: WooCommerce/Classes/ViewRelated/Orders/OrderStatusListViewController.swift [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        registerTableViewCells()
+        configureNavigationBar()
+        configureTableView()
+
+        configureResultsController()
+    }
+
```
Comment (mindgraffiti) [inline WooCommerce/Classes/ViewRelated/Orders/OrderStatusListViewController.swift]: Let's add some Woo styling here. Change the status bar to white.

```
override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        navigationController?.navigationBar.barStyle = .black 
    }
```

File: WooCommerce/Classes/ViewRelated/Orders/OrderStatusListViewController.swift [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+    @objc func applyButtonTapped() {
+        updateOrderStatus()
+        dismiss(animated: true, completion: nil)
+    }
+}
+
+
+/// MARK: - Update the Order status
+///
```
Comment (bummytime) [inline WooCommerce/Classes/ViewRelated/Orders/OrderStatusListViewController.swift]: Just a style nit — you may want to change the `/// MARK: - ...` comments to 

```
// MARK: - Update the Order status
//
```

So we get the nice seporators in the XCode pull-down:

screen shot 2019-03-07 at 12 10 10 pm

File: WooCommerce/WooCommerceTests/Views/StatusListTableViewCellTests.swift [hunk 5]
```diff
@@ -0,0 +1,34 @@
+import XCTest
+@testable import WooCommerce
+
+final class StatusListTableViewCellTests: XCTestCase {
```
Comment (bummytime) [inline WooCommerce/WooCommerceTests/Views/StatusListTableViewCellTests.swift]: Thank you for the UI unit tests! 🙇🙇🙇

File: WooCommerce/Classes/ViewRelated/Orders/OrderStatusListViewController.swift [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+        }
+    }
+
+    /// Setup: TableView
+    ///
+    func configureTableView() {
+        view.backgroundColor = StyleManager.tableViewBackgroundColor
+        tableView.backgroundColor = StyleManager.tableViewBackgroundColor
+        tableView.refreshControl = refreshControl
+
```
Comment (mindgraffiti) [inline WooCommerce/Classes/ViewRelated/Orders/OrderStatusListViewController.swift]: More Woo styling: don't forget to change the tableview in the .xib file to be a `grouped` table style.

File: WooCommerce/Classes/ViewRelated/Orders/OrderStatusListViewController.swift [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+            DDLogError("⛔️ Order Update Failure: [\(orderID).status = \(statusKey)]. Error: \(error)")
+
+            self.displayErrorNotice(orderID: orderID)
+        })
+    }
+}
+
+
+/// MARK: - UITableViewDatasource conformance
+extension OrderStatusListViewController: UITableViewDataSource {
```
Comment (mindgraffiti) [inline WooCommerce/Classes/ViewRelated/Orders/OrderStatusListViewController.swift]: To collapse the section title on a grouped table, add these delegate methods:

```
    func numberOfSections(in tableView: UITableView) -> Int {
        return statusResultsController.sections.count
    }

    func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        // iOS 11 table bug. Must return a tiny value to collapse `nil` or `empty` section headers.
        return CGFloat.leastNonzeroMagnitude
    }
```

File: WooCommerce/Classes/ViewRelated/Orders/OrderStatusListViewController.swift [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+extension OrderStatusListViewController {
+    /// Displays the `Order Updated` Notice. Whenever the `Undo` button gets pressed, we'll execute the `onUndoAction` closure.
+    ///
+    private func displayOrderUpdatedNotice(onUndoAction: @escaping () -> Void) {
+        let message = NSLocalizedString("Order status updated", comment: "Order status update success notice")
+        let actionTitle = NSLocalizedString("Undo", comment: "Undo Action")
+        let notice = Notice(title: message, feedbackType: .success, actionTitle: actionTitle, actionHandler: onUndoAction)
+
+        AppDelegate.shared.noticePresenter.enqueue(notice: notice)
+    }
```
Comment (mindgraffiti) [inline WooCommerce/Classes/ViewRelated/Orders/OrderStatusListViewController.swift]: Woo code styling tip: add some breathing room to your code. Leave a newline between an ending curly brace and a new comment.

```
}

/// Comment
///
func display()
```

**Comment by ctarda** [comment]
Thanks a lot for your comments @bummytime and @mindgraffiti ! 

I have pushed an update with your feedback, will merge as soon as the build passes

**Pull request merged**